{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2a3ffea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e36f42a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the maximum number of columns/rows displayed in the cell\n",
    "#pd.reset_option('display.max_columns', 20)\n",
    "#pd.set_option('display.max_rows', 20)\n",
    "#pd.set_option('display.max_columns', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "604de76d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_TPM(expression_file_path, fold):\n",
    "    df = pd.read_csv(expression_file_path, sep='\\t')\n",
    "    \n",
    "    df['TPM_ut_mean'] = (df['TPM'] + df['TPM.1'] + df['TPM.2']) / 3\n",
    "    df['TPM_ef_mean'] = (df['TPM.6'] + df['TPM.7'] + df['TPM.8']) / 3\n",
    "\n",
    "    # filter transcripts that have TPM > 1\n",
    "    df_gt1 = df[df['TPM_ef_mean'] > 1]\n",
    "\n",
    "    df_gt1['fold_change'] = df_gt1['TPM_ef_mean'] / df_gt1['TPM_ut_mean']\n",
    "\n",
    "    # filter transcripts that have TPM fold change ef/ut > fold (= 50)\n",
    "    high_fold_gene = df_gt1[df_gt1['fold_change'] > fold].sort_values(by='fold_change', ascending=False)\n",
    "    \n",
    "    return(high_fold_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bae723ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_gtf(gtf_file_path):\n",
    "    # read gtf file into dataframe gtf\n",
    "    column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "    gtf = pd.read_csv(gtf_file_path, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)\n",
    "    \n",
    "    # label if each entry has reference gene id\n",
    "    for i in range(gtf.shape[0]):\n",
    "        if 'ref_gene_id' in gtf.loc[i,'attribute']:\n",
    "            gtf.loc[i,'if_ref_id'] = 'T'\n",
    "        else:\n",
    "            gtf.loc[i,'if_ref_id'] = 'F'\n",
    "    \n",
    "    # extract transcripts without ref_id\n",
    "    new_transcript = gtf[(gtf['feature'] == 'transcript') & (gtf['if_ref_id'] == 'F')]\n",
    "    \n",
    "    return new_transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "abdcda89",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_id(input_string):\n",
    "    # regular expression to catch the content in the quotes after 'gene_id'\n",
    "    pattern = r'gene_id\\s+\"([^\"]+)\"'\n",
    "    \n",
    "    # try to find the pattern\n",
    "    match = re.search(pattern, input_string)\n",
    "\n",
    "    # if find the pattern, return the content in the quotes\n",
    "    if match:\n",
    "        return(match.group(1))\n",
    "    return 'None'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "46fe873c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def screen_data(high_fold_gene, new_transcript):\n",
    "    # create a new dataframe for screened data\n",
    "    df_new = pd.DataFrame(columns=high_fold_gene.columns)\n",
    "\n",
    "    # check every high fold-change gene if it's contained in new_transcript list\n",
    "    for i in high_fold_gene.index:\n",
    "        for j in new_transcript.index[:]:\n",
    "            if high_fold_gene.loc[i,'Gene ID'] == get_gene_id(new_transcript.loc[j,'attribute']):\n",
    "                # add the line containg gene info to the dataframe\n",
    "                df_new = pd.concat([df_new, high_fold_gene.loc[i, :].to_frame().transpose()], ignore_index=True)\n",
    "\n",
    "    # remove the replicates\n",
    "    df_new = df_new.drop_duplicates()\n",
    "    \n",
    "    return(df_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "48ad7f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run(expression_file_path, gtf_file_path, fold):\n",
    "    high_fold_gene = filter_TPM(expression_file_path, fold)\n",
    "    new_transcript = filter_gtf(gtf_file_path)\n",
    "    df_new = screen_data(high_fold_gene, new_transcript)\n",
    "    \n",
    "    return df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9cc3268e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_84884/3275195340.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_gt1['fold_change'] = df_gt1['TPM_ef_mean'] / df_gt1['TPM_ut_mean']\n"
     ]
    }
   ],
   "source": [
    "expression_file_path = 'ef_merged_expression/Dmel_comb.txt'\n",
    "gtf_file_path = 'ef_merged_gtf/Dmel_merged.gtf'\n",
    "df_dmel = run(expression_file_path, gtf_file_path, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2e059f76-90e8-41e4-9924-b466e490419e",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "gtf = pd.read_csv(gtf_file_path, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2b877a1d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'TPM_ut1'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/envs/research/lib/python3.8/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3620\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3621\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3622\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/research/lib/python3.8/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/research/lib/python3.8/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'TPM_ut1'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_77462/1003865465.py\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mexpression_file_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'ef_merged_expression/Sleb_comb.txt'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mgtf_file_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'ef_merged_gtf/Sleb_merged.gtf'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mdf_sleb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpression_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgtf_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_77462/3243286875.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(expression_file_path, gtf_file_path, fold)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpression_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgtf_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfold\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mhigh_fold_gene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfilter_TPM\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpression_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfold\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mnew_transcript\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfilter_gtf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgtf_file_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mdf_new\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscreen_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhigh_fold_gene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_transcript\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_77462/1289949573.py\u001b[0m in \u001b[0;36mfilter_TPM\u001b[0;34m(expression_file_path, fold)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexpression_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ut_mean'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ut1'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ut2'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ut3'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ef_mean'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ef1'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ef2'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TPM_ef3'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/research/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3503\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3504\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3505\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3506\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3507\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/research/lib/python3.8/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3621\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3622\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3623\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3624\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3625\u001b[0m                 \u001b[0;31m# If we have a listlike key, _check_indexing_error will raise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'TPM_ut1'"
     ]
    }
   ],
   "source": [
    "expression_file_path = 'ef_merged_expression/Sleb_comb.txt'\n",
    "gtf_file_path = 'ef_merged_gtf/Sleb_merged.gtf'\n",
    "df_sleb = run(expression_file_path, gtf_file_path, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1bd80172",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function extracts transcript seuqences of genes of interest from a fasta file containing all transcripts\n",
    "# input is transcript sequences file path, a gene list, and output file path\n",
    "def extract_sequences(transcript_file_path, gene_list, output_file_path):\n",
    "    \n",
    "    sequences = []\n",
    "\n",
    "    for record in SeqIO.parse(transcript_file_path, \"fasta\"):\n",
    "        # convert transcript_id to gene_id by removing the content after last '.'\n",
    "        parts = record.id.rsplit('.', 1)\n",
    "        gene_id = parts[0] if len(parts) > 1 else record.id\n",
    "\n",
    "        if gene_id in gene_list:\n",
    "            sequences.append(record)\n",
    "\n",
    "    with open(output_file_path, \"w\") as output_handle:\n",
    "        SeqIO.write(sequences, output_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "655f2447",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_file_path = 'sleb_m_transcript.fa'\n",
    "output_file_path = 'sleb_new_trasncript_seq.fasta'\n",
    "extract_sequences(transcript_file_path, df_filter_list, output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e582630c-8e3c-41bb-8fab-df683625c3e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file_path = 'GCF_003285725.1_SlebRS2_genomic.gff'\n",
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "df = pd.read_csv(gtf_file_path, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)\n",
    "\n",
    "ref_gtf = df[df['feature']=='exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1dfbdd10-fe6c-47ee-8b87-48becbe07cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if two features(exons/transcripts/genes) have overlapping regions by comparing their start/end locations\n",
    "# return True if overlapped\n",
    "def check_overlap(feature1, feature2):\n",
    "    return feature1['Start'] <= feature2['end'] and feature1['End'] >= feature2['start']\n",
    "\n",
    "def add_overlap_info(df_in,ref_gtf):\n",
    "    output = pd.DataFrame()\n",
    "    # create a dictionary saving the gtf for each contig\n",
    "    ref_gtf_contig = {}\n",
    "    \n",
    "    for i in df_in.index:\n",
    "        df_in.loc[i,'if_overlap'] = 'No'\n",
    "        contig = df_in.loc[i,'Reference']\n",
    "        \n",
    "        # each key is contig name, each value the gtf info for the contig\n",
    "        if contig not in ref_gtf_contig.keys():\n",
    "            ref_gtf_contig[contig] = ref_gtf[ref_gtf['seqname'] == contig].sort_values(by='start', ascending=True)\n",
    "        for j in ref_gtf_contig[contig].index:\n",
    "            if ref_gtf.loc[j,'start'] > df_in.loc[i,'End']:\n",
    "                break\n",
    "            # check if the transcript has any overlap regions with exisiting exons\n",
    "            if check_overlap(df_in.loc[i,:],ref_gtf.loc[j,:]):\n",
    "                start = max(df_in.loc[i,'Start'],ref_gtf.loc[j,'start'])\n",
    "                end = min(df_in.loc[i,'End'],ref_gtf.loc[j,'end'])\n",
    "                overlap_length = end - start\n",
    "                length = df_in.loc[i,'End'] - df_in.loc[i,'Start']\n",
    "                percent = overlap_length / length\n",
    "\n",
    "                # if it's the first overlap region found, update it\n",
    "                if df_in.loc[i,'if_overlap'] == 'No':\n",
    "                    df_in.loc[i,'if_overlap'] = 'Yes'\n",
    "                    df_in.loc[i,'%overlap'] = percent\n",
    "                # if not the first, update with the larger overlap region\n",
    "                else:\n",
    "                    if df_in.loc[i,'%overlap'] < percent:\n",
    "                        df_in.loc[i,'%overlap'] = percent\n",
    "    return df_in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b64352a8-b446-4ce1-8c70-d9a864d448cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_sleb = df_sleb.loc[df_filter.index,:]\n",
    "test_sleb = test_sleb.sort_values(by='Start', ascending=True)\n",
    "added_df_in = add_overlap_info(test_sleb, ref_gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "7b073cbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM.1</th>\n",
       "      <th>...</th>\n",
       "      <th>TPM.6</th>\n",
       "      <th>FPKM.7</th>\n",
       "      <th>TPM.7</th>\n",
       "      <th>FPKM.8</th>\n",
       "      <th>TPM.8</th>\n",
       "      <th>TPM_ut_mean</th>\n",
       "      <th>TPM_pr_mean</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>if_overlap</th>\n",
       "      <th>%overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>MSTRG.6046</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060550.1</td>\n",
       "      <td>.</td>\n",
       "      <td>61228</td>\n",
       "      <td>61556</td>\n",
       "      <td>5.048632</td>\n",
       "      <td>1.757874</td>\n",
       "      <td>3.506197</td>\n",
       "      <td>1.413304</td>\n",
       "      <td>...</td>\n",
       "      <td>8.452106</td>\n",
       "      <td>6.702477</td>\n",
       "      <td>13.094912</td>\n",
       "      <td>5.676188</td>\n",
       "      <td>11.449466</td>\n",
       "      <td>3.861828</td>\n",
       "      <td>10.998828</td>\n",
       "      <td>2.848088</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>MSTRG.12928</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060742.1</td>\n",
       "      <td>+</td>\n",
       "      <td>232477</td>\n",
       "      <td>236036</td>\n",
       "      <td>7.945543</td>\n",
       "      <td>4.649685</td>\n",
       "      <td>9.274108</td>\n",
       "      <td>4.451277</td>\n",
       "      <td>...</td>\n",
       "      <td>15.114046</td>\n",
       "      <td>10.657976</td>\n",
       "      <td>20.822937</td>\n",
       "      <td>9.418137</td>\n",
       "      <td>18.99737</td>\n",
       "      <td>8.5465</td>\n",
       "      <td>18.311451</td>\n",
       "      <td>2.142567</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.193875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>MSTRG.11231</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060621.1</td>\n",
       "      <td>-</td>\n",
       "      <td>303865</td>\n",
       "      <td>306706</td>\n",
       "      <td>2.167556</td>\n",
       "      <td>0.786978</td>\n",
       "      <td>1.56968</td>\n",
       "      <td>2.602611</td>\n",
       "      <td>...</td>\n",
       "      <td>7.628936</td>\n",
       "      <td>2.818815</td>\n",
       "      <td>5.507238</td>\n",
       "      <td>2.821084</td>\n",
       "      <td>5.690422</td>\n",
       "      <td>3.103453</td>\n",
       "      <td>6.275532</td>\n",
       "      <td>2.022113</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>MSTRG.2502</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060538.1</td>\n",
       "      <td>.</td>\n",
       "      <td>354510</td>\n",
       "      <td>355653</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.053567</td>\n",
       "      <td>0.106844</td>\n",
       "      <td>3.656264</td>\n",
       "      <td>...</td>\n",
       "      <td>13.107974</td>\n",
       "      <td>2.038422</td>\n",
       "      <td>3.98255</td>\n",
       "      <td>2.926721</td>\n",
       "      <td>5.903502</td>\n",
       "      <td>2.925438</td>\n",
       "      <td>7.664675</td>\n",
       "      <td>2.620009</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.11019</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060600.1</td>\n",
       "      <td>.</td>\n",
       "      <td>566404</td>\n",
       "      <td>566732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>47.595459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.865153</td>\n",
       "      <td>inf</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>MSTRG.1588</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>.</td>\n",
       "      <td>18391518</td>\n",
       "      <td>18391821</td>\n",
       "      <td>18.536184</td>\n",
       "      <td>6.45408</td>\n",
       "      <td>12.873096</td>\n",
       "      <td>144.664841</td>\n",
       "      <td>...</td>\n",
       "      <td>6260.32373</td>\n",
       "      <td>3970.977783</td>\n",
       "      <td>7758.266113</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.576414</td>\n",
       "      <td>4672.863281</td>\n",
       "      <td>47.88927</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>MSTRG.1943</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>-</td>\n",
       "      <td>22186799</td>\n",
       "      <td>22217630</td>\n",
       "      <td>11.086722</td>\n",
       "      <td>7.554018</td>\n",
       "      <td>15.066996</td>\n",
       "      <td>8.427059</td>\n",
       "      <td>...</td>\n",
       "      <td>28.848961</td>\n",
       "      <td>27.170868</td>\n",
       "      <td>53.084869</td>\n",
       "      <td>15.257482</td>\n",
       "      <td>30.775938</td>\n",
       "      <td>15.239915</td>\n",
       "      <td>37.569923</td>\n",
       "      <td>2.465232</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.052155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>MSTRG.4005</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060538.1</td>\n",
       "      <td>+</td>\n",
       "      <td>23033946</td>\n",
       "      <td>23036132</td>\n",
       "      <td>1.401538</td>\n",
       "      <td>0.532584</td>\n",
       "      <td>1.062275</td>\n",
       "      <td>0.481346</td>\n",
       "      <td>...</td>\n",
       "      <td>4.362121</td>\n",
       "      <td>6.606958</td>\n",
       "      <td>12.908292</td>\n",
       "      <td>3.466725</td>\n",
       "      <td>6.992747</td>\n",
       "      <td>0.918917</td>\n",
       "      <td>8.08772</td>\n",
       "      <td>8.801361</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.232845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MSTRG.2269</td>\n",
       "      <td>.</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>.</td>\n",
       "      <td>27209526</td>\n",
       "      <td>27209781</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.352944</td>\n",
       "      <td>...</td>\n",
       "      <td>1058.042847</td>\n",
       "      <td>750.417053</td>\n",
       "      <td>1466.12146</td>\n",
       "      <td>550.123291</td>\n",
       "      <td>1109.656372</td>\n",
       "      <td>4.774623</td>\n",
       "      <td>1211.27356</td>\n",
       "      <td>253.689867</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>MSTRG.2345</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>.</td>\n",
       "      <td>28632068</td>\n",
       "      <td>28632714</td>\n",
       "      <td>20.98609</td>\n",
       "      <td>7.307108</td>\n",
       "      <td>14.574517</td>\n",
       "      <td>4.962482</td>\n",
       "      <td>...</td>\n",
       "      <td>23.659952</td>\n",
       "      <td>9.761988</td>\n",
       "      <td>19.072409</td>\n",
       "      <td>15.184164</td>\n",
       "      <td>30.62805</td>\n",
       "      <td>10.986646</td>\n",
       "      <td>24.45347</td>\n",
       "      <td>2.225745</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene ID Gene Name       Reference Strand     Start       End  \\\n",
       "291   MSTRG.6046         -  NW_022060550.1      .     61228     61556   \n",
       "468  MSTRG.12928         -  NW_022060742.1      +    232477    236036   \n",
       "508  MSTRG.11231         -  NW_022060621.1      -    303865    306706   \n",
       "328   MSTRG.2502         -  NW_022060538.1      .    354510    355653   \n",
       "3    MSTRG.11019         -  NW_022060600.1      .    566404    566732   \n",
       "..           ...       ...             ...    ...       ...       ...   \n",
       "70    MSTRG.1588         -  NW_022060529.1      .  18391518  18391821   \n",
       "361   MSTRG.1943         -  NW_022060529.1      -  22186799  22217630   \n",
       "106   MSTRG.4005         -  NW_022060538.1      +  23033946  23036132   \n",
       "61    MSTRG.2269         .  NW_022060529.1      .  27209526  27209781   \n",
       "414   MSTRG.2345         -  NW_022060529.1      .  28632068  28632714   \n",
       "\n",
       "      Coverage      FPKM        TPM      FPKM.1  ...        TPM.6  \\\n",
       "291   5.048632  1.757874   3.506197    1.413304  ...     8.452106   \n",
       "468   7.945543  4.649685   9.274108    4.451277  ...    15.114046   \n",
       "508   2.167556  0.786978    1.56968    2.602611  ...     7.628936   \n",
       "328   0.153846  0.053567   0.106844    3.656264  ...    13.107974   \n",
       "3          0.0       0.0        0.0         0.0  ...    47.595459   \n",
       "..         ...       ...        ...         ...  ...          ...   \n",
       "70   18.536184   6.45408  12.873096  144.664841  ...   6260.32373   \n",
       "361  11.086722  7.554018  15.066996    8.427059  ...    28.848961   \n",
       "106   1.401538  0.532584   1.062275    0.481346  ...     4.362121   \n",
       "61         0.0       0.0        0.0    7.352944  ...  1058.042847   \n",
       "414   20.98609  7.307108  14.574517    4.962482  ...    23.659952   \n",
       "\n",
       "          FPKM.7        TPM.7      FPKM.8        TPM.8 TPM_ut_mean  \\\n",
       "291     6.702477    13.094912    5.676188    11.449466    3.861828   \n",
       "468    10.657976    20.822937    9.418137     18.99737      8.5465   \n",
       "508     2.818815     5.507238    2.821084     5.690422    3.103453   \n",
       "328     2.038422      3.98255    2.926721     5.903502    2.925438   \n",
       "3            0.0          0.0         0.0          0.0         0.0   \n",
       "..           ...          ...         ...          ...         ...   \n",
       "70   3970.977783  7758.266113         0.0          0.0   97.576414   \n",
       "361    27.170868    53.084869   15.257482    30.775938   15.239915   \n",
       "106     6.606958    12.908292    3.466725     6.992747    0.918917   \n",
       "61    750.417053   1466.12146  550.123291  1109.656372    4.774623   \n",
       "414     9.761988    19.072409   15.184164     30.62805   10.986646   \n",
       "\n",
       "     TPM_pr_mean fold_change if_overlap  %overlap  \n",
       "291    10.998828    2.848088         No       NaN  \n",
       "468    18.311451    2.142567        Yes  0.193875  \n",
       "508     6.275532    2.022113         No       NaN  \n",
       "328     7.664675    2.620009         No       NaN  \n",
       "3      15.865153         inf         No       NaN  \n",
       "..           ...         ...        ...       ...  \n",
       "70   4672.863281    47.88927         No       NaN  \n",
       "361    37.569923    2.465232        Yes  0.052155  \n",
       "106      8.08772    8.801361        Yes  0.232845  \n",
       "61    1211.27356  253.689867         No       NaN  \n",
       "414     24.45347    2.225745         No       NaN  \n",
       "\n",
       "[62 rows x 30 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "added_df_in[(added_df_in['if_overlap'] != 'Yes') | (added_df_in['%overlap'] < 0.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "de8e4c26",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM.1</th>\n",
       "      <th>...</th>\n",
       "      <th>TPM.6</th>\n",
       "      <th>FPKM.7</th>\n",
       "      <th>TPM.7</th>\n",
       "      <th>FPKM.8</th>\n",
       "      <th>TPM.8</th>\n",
       "      <th>TPM_ut_mean</th>\n",
       "      <th>TPM_pr_mean</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>if_overlap</th>\n",
       "      <th>%overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.11019</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060600.1</td>\n",
       "      <td>.</td>\n",
       "      <td>566404</td>\n",
       "      <td>566732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>47.595459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.865153</td>\n",
       "      <td>inf</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MSTRG.8439</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060566.1</td>\n",
       "      <td>.</td>\n",
       "      <td>2721671</td>\n",
       "      <td>2722513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.188394</td>\n",
       "      <td>3.751498</td>\n",
       "      <td>7.32946</td>\n",
       "      <td>3.622036</td>\n",
       "      <td>7.306027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.60796</td>\n",
       "      <td>inf</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MSTRG.2269</td>\n",
       "      <td>.</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>.</td>\n",
       "      <td>27209526</td>\n",
       "      <td>27209781</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.352944</td>\n",
       "      <td>...</td>\n",
       "      <td>1058.042847</td>\n",
       "      <td>750.417053</td>\n",
       "      <td>1466.12146</td>\n",
       "      <td>550.123291</td>\n",
       "      <td>1109.656372</td>\n",
       "      <td>4.774623</td>\n",
       "      <td>1211.27356</td>\n",
       "      <td>253.689867</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>MSTRG.1588</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>.</td>\n",
       "      <td>18391518</td>\n",
       "      <td>18391821</td>\n",
       "      <td>18.536184</td>\n",
       "      <td>6.45408</td>\n",
       "      <td>12.873096</td>\n",
       "      <td>144.664841</td>\n",
       "      <td>...</td>\n",
       "      <td>6260.32373</td>\n",
       "      <td>3970.977783</td>\n",
       "      <td>7758.266113</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.576414</td>\n",
       "      <td>4672.863281</td>\n",
       "      <td>47.88927</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>MSTRG.4616</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060545.1</td>\n",
       "      <td>.</td>\n",
       "      <td>7870440</td>\n",
       "      <td>7870931</td>\n",
       "      <td>1.888211</td>\n",
       "      <td>0.657453</td>\n",
       "      <td>1.311334</td>\n",
       "      <td>45.388313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>688.548279</td>\n",
       "      <td>1388.874146</td>\n",
       "      <td>29.122767</td>\n",
       "      <td>462.958049</td>\n",
       "      <td>15.896774</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>MSTRG.2516</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060538.1</td>\n",
       "      <td>.</td>\n",
       "      <td>655049</td>\n",
       "      <td>655722</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>97.198975</td>\n",
       "      <td>49.425846</td>\n",
       "      <td>96.565361</td>\n",
       "      <td>41.819115</td>\n",
       "      <td>84.353539</td>\n",
       "      <td>24.563093</td>\n",
       "      <td>92.705958</td>\n",
       "      <td>3.774197</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060742.1</td>\n",
       "      <td>-</td>\n",
       "      <td>12738711</td>\n",
       "      <td>12750714</td>\n",
       "      <td>2.963251</td>\n",
       "      <td>1.031769</td>\n",
       "      <td>2.057932</td>\n",
       "      <td>1.329909</td>\n",
       "      <td>...</td>\n",
       "      <td>5.184815</td>\n",
       "      <td>3.801761</td>\n",
       "      <td>7.42766</td>\n",
       "      <td>3.653596</td>\n",
       "      <td>7.369686</td>\n",
       "      <td>1.833778</td>\n",
       "      <td>6.66072</td>\n",
       "      <td>3.632239</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>MSTRG.11598</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060738.1</td>\n",
       "      <td>+</td>\n",
       "      <td>1820405</td>\n",
       "      <td>1820902</td>\n",
       "      <td>4.790068</td>\n",
       "      <td>1.667845</td>\n",
       "      <td>3.326629</td>\n",
       "      <td>2.613643</td>\n",
       "      <td>...</td>\n",
       "      <td>18.664377</td>\n",
       "      <td>7.19315</td>\n",
       "      <td>14.05356</td>\n",
       "      <td>3.428897</td>\n",
       "      <td>6.916446</td>\n",
       "      <td>4.003816</td>\n",
       "      <td>13.211461</td>\n",
       "      <td>3.299717</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>MSTRG.6046</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060550.1</td>\n",
       "      <td>.</td>\n",
       "      <td>61228</td>\n",
       "      <td>61556</td>\n",
       "      <td>5.048632</td>\n",
       "      <td>1.757874</td>\n",
       "      <td>3.506197</td>\n",
       "      <td>1.413304</td>\n",
       "      <td>...</td>\n",
       "      <td>8.452106</td>\n",
       "      <td>6.702477</td>\n",
       "      <td>13.094912</td>\n",
       "      <td>5.676188</td>\n",
       "      <td>11.449466</td>\n",
       "      <td>3.861828</td>\n",
       "      <td>10.998828</td>\n",
       "      <td>2.848088</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>MSTRG.7930</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060564.1</td>\n",
       "      <td>.</td>\n",
       "      <td>5152968</td>\n",
       "      <td>5153611</td>\n",
       "      <td>11.07764</td>\n",
       "      <td>3.857103</td>\n",
       "      <td>7.693251</td>\n",
       "      <td>6.31877</td>\n",
       "      <td>...</td>\n",
       "      <td>32.050354</td>\n",
       "      <td>15.168712</td>\n",
       "      <td>29.635752</td>\n",
       "      <td>12.087071</td>\n",
       "      <td>24.38089</td>\n",
       "      <td>10.853178</td>\n",
       "      <td>28.688999</td>\n",
       "      <td>2.643373</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>MSTRG.3111</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060538.1</td>\n",
       "      <td>-</td>\n",
       "      <td>10357609</td>\n",
       "      <td>10358156</td>\n",
       "      <td>3.320236</td>\n",
       "      <td>1.198206</td>\n",
       "      <td>2.389902</td>\n",
       "      <td>1.680997</td>\n",
       "      <td>...</td>\n",
       "      <td>8.626204</td>\n",
       "      <td>3.754206</td>\n",
       "      <td>7.33475</td>\n",
       "      <td>2.25972</td>\n",
       "      <td>4.558091</td>\n",
       "      <td>2.602278</td>\n",
       "      <td>6.839682</td>\n",
       "      <td>2.628344</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>MSTRG.2502</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060538.1</td>\n",
       "      <td>.</td>\n",
       "      <td>354510</td>\n",
       "      <td>355653</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.053567</td>\n",
       "      <td>0.106844</td>\n",
       "      <td>3.656264</td>\n",
       "      <td>...</td>\n",
       "      <td>13.107974</td>\n",
       "      <td>2.038422</td>\n",
       "      <td>3.98255</td>\n",
       "      <td>2.926721</td>\n",
       "      <td>5.903502</td>\n",
       "      <td>2.925438</td>\n",
       "      <td>7.664675</td>\n",
       "      <td>2.620009</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>+</td>\n",
       "      <td>5277255</td>\n",
       "      <td>5279891</td>\n",
       "      <td>19.634945</td>\n",
       "      <td>6.836656</td>\n",
       "      <td>13.636168</td>\n",
       "      <td>4.972131</td>\n",
       "      <td>...</td>\n",
       "      <td>30.243635</td>\n",
       "      <td>16.090231</td>\n",
       "      <td>31.436163</td>\n",
       "      <td>12.162856</td>\n",
       "      <td>24.533756</td>\n",
       "      <td>11.896074</td>\n",
       "      <td>28.737851</td>\n",
       "      <td>2.415742</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>MSTRG.2345</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060529.1</td>\n",
       "      <td>.</td>\n",
       "      <td>28632068</td>\n",
       "      <td>28632714</td>\n",
       "      <td>20.98609</td>\n",
       "      <td>7.307108</td>\n",
       "      <td>14.574517</td>\n",
       "      <td>4.962482</td>\n",
       "      <td>...</td>\n",
       "      <td>23.659952</td>\n",
       "      <td>9.761988</td>\n",
       "      <td>19.072409</td>\n",
       "      <td>15.184164</td>\n",
       "      <td>30.62805</td>\n",
       "      <td>10.986646</td>\n",
       "      <td>24.45347</td>\n",
       "      <td>2.225745</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>MSTRG.6108</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060554.1</td>\n",
       "      <td>.</td>\n",
       "      <td>1331725</td>\n",
       "      <td>1332680</td>\n",
       "      <td>6.292887</td>\n",
       "      <td>2.191109</td>\n",
       "      <td>4.370314</td>\n",
       "      <td>1.931814</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643511</td>\n",
       "      <td>2.74434</td>\n",
       "      <td>5.361732</td>\n",
       "      <td>4.064538</td>\n",
       "      <td>8.198601</td>\n",
       "      <td>3.384568</td>\n",
       "      <td>7.401281</td>\n",
       "      <td>2.186773</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>MSTRG.11231</td>\n",
       "      <td>-</td>\n",
       "      <td>NW_022060621.1</td>\n",
       "      <td>-</td>\n",
       "      <td>303865</td>\n",
       "      <td>306706</td>\n",
       "      <td>2.167556</td>\n",
       "      <td>0.786978</td>\n",
       "      <td>1.56968</td>\n",
       "      <td>2.602611</td>\n",
       "      <td>...</td>\n",
       "      <td>7.628936</td>\n",
       "      <td>2.818815</td>\n",
       "      <td>5.507238</td>\n",
       "      <td>2.821084</td>\n",
       "      <td>5.690422</td>\n",
       "      <td>3.103453</td>\n",
       "      <td>6.275532</td>\n",
       "      <td>2.022113</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene ID Gene Name       Reference Strand     Start       End  \\\n",
       "3    MSTRG.11019         -  NW_022060600.1      .    566404    566732   \n",
       "33    MSTRG.8439         -  NW_022060566.1      .   2721671   2722513   \n",
       "61    MSTRG.2269         .  NW_022060529.1      .  27209526  27209781   \n",
       "70    MSTRG.1588         -  NW_022060529.1      .  18391518  18391821   \n",
       "88    MSTRG.4616         -  NW_022060545.1      .   7870440   7870931   \n",
       "220   MSTRG.2516         -  NW_022060538.1      .    655049    655722   \n",
       "228  MSTRG.13746         -  NW_022060742.1      -  12738711  12750714   \n",
       "252  MSTRG.11598         -  NW_022060738.1      +   1820405   1820902   \n",
       "291   MSTRG.6046         -  NW_022060550.1      .     61228     61556   \n",
       "323   MSTRG.7930         -  NW_022060564.1      .   5152968   5153611   \n",
       "325   MSTRG.3111         -  NW_022060538.1      -  10357609  10358156   \n",
       "328   MSTRG.2502         -  NW_022060538.1      .    354510    355653   \n",
       "377    MSTRG.406         -  NW_022060529.1      +   5277255   5279891   \n",
       "414   MSTRG.2345         -  NW_022060529.1      .  28632068  28632714   \n",
       "443   MSTRG.6108         -  NW_022060554.1      .   1331725   1332680   \n",
       "508  MSTRG.11231         -  NW_022060621.1      -    303865    306706   \n",
       "\n",
       "      Coverage      FPKM        TPM      FPKM.1  ...        TPM.6  \\\n",
       "3          0.0       0.0        0.0         0.0  ...    47.595459   \n",
       "33         0.0       0.0        0.0         0.0  ...     5.188394   \n",
       "61         0.0       0.0        0.0    7.352944  ...  1058.042847   \n",
       "70   18.536184   6.45408  12.873096  144.664841  ...   6260.32373   \n",
       "88    1.888211  0.657453   1.311334   45.388313  ...          0.0   \n",
       "220        0.0       0.0        0.0         0.0  ...    97.198975   \n",
       "228   2.963251  1.031769   2.057932    1.329909  ...     5.184815   \n",
       "252   4.790068  1.667845   3.326629    2.613643  ...    18.664377   \n",
       "291   5.048632  1.757874   3.506197    1.413304  ...     8.452106   \n",
       "323   11.07764  3.857103   7.693251     6.31877  ...    32.050354   \n",
       "325   3.320236  1.198206   2.389902    1.680997  ...     8.626204   \n",
       "328   0.153846  0.053567   0.106844    3.656264  ...    13.107974   \n",
       "377  19.634945  6.836656  13.636168    4.972131  ...    30.243635   \n",
       "414   20.98609  7.307108  14.574517    4.962482  ...    23.659952   \n",
       "443   6.292887  2.191109   4.370314    1.931814  ...     8.643511   \n",
       "508   2.167556  0.786978    1.56968    2.602611  ...     7.628936   \n",
       "\n",
       "          FPKM.7        TPM.7      FPKM.8        TPM.8 TPM_ut_mean  \\\n",
       "3            0.0          0.0         0.0          0.0         0.0   \n",
       "33      3.751498      7.32946    3.622036     7.306027         0.0   \n",
       "61    750.417053   1466.12146  550.123291  1109.656372    4.774623   \n",
       "70   3970.977783  7758.266113         0.0          0.0   97.576414   \n",
       "88           0.0          0.0  688.548279  1388.874146   29.122767   \n",
       "220    49.425846    96.565361   41.819115    84.353539   24.563093   \n",
       "228     3.801761      7.42766    3.653596     7.369686    1.833778   \n",
       "252      7.19315     14.05356    3.428897     6.916446    4.003816   \n",
       "291     6.702477    13.094912    5.676188    11.449466    3.861828   \n",
       "323    15.168712    29.635752   12.087071     24.38089   10.853178   \n",
       "325     3.754206      7.33475     2.25972     4.558091    2.602278   \n",
       "328     2.038422      3.98255    2.926721     5.903502    2.925438   \n",
       "377    16.090231    31.436163   12.162856    24.533756   11.896074   \n",
       "414     9.761988    19.072409   15.184164     30.62805   10.986646   \n",
       "443      2.74434     5.361732    4.064538     8.198601    3.384568   \n",
       "508     2.818815     5.507238    2.821084     5.690422    3.103453   \n",
       "\n",
       "     TPM_pr_mean fold_change if_overlap %overlap  \n",
       "3      15.865153         inf         No      NaN  \n",
       "33       6.60796         inf         No      NaN  \n",
       "61    1211.27356  253.689867         No      NaN  \n",
       "70   4672.863281    47.88927         No      NaN  \n",
       "88    462.958049   15.896774         No      NaN  \n",
       "220    92.705958    3.774197         No      NaN  \n",
       "228      6.66072    3.632239         No      NaN  \n",
       "252    13.211461    3.299717         No      NaN  \n",
       "291    10.998828    2.848088         No      NaN  \n",
       "323    28.688999    2.643373         No      NaN  \n",
       "325     6.839682    2.628344         No      NaN  \n",
       "328     7.664675    2.620009         No      NaN  \n",
       "377    28.737851    2.415742         No      NaN  \n",
       "414     24.45347    2.225745         No      NaN  \n",
       "443     7.401281    2.186773         No      NaN  \n",
       "508     6.275532    2.022113         No      NaN  \n",
       "\n",
       "[16 rows x 30 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "added_df_in[added_df_in['if_overlap'] == 'No'].sort_values(by='fold_change', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "26022f71-594e-48a2-acb0-da6a3a38403b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_41839/3275195340.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_gt1['fold_change'] = df_gt1['TPM_ef_mean'] / df_gt1['TPM_ut_mean']\n"
     ]
    }
   ],
   "source": [
    "## Sleb\n",
    "expression_file_path_sleb = 'ef_merged_expression/Sleb_comb.txt'\n",
    "gtf_file_path_sleb = 'ef_merged_gtf/Sleb_merged.gtf'\n",
    "df_sleb = run(expression_file_path_sleb, gtf_file_path_sleb, 2)\n",
    "\n",
    "# filter for foldchange > 2, meanTPM of  > 5, and don't contain LOC ref gene_id\n",
    "# .iloc[:,[0,4,5,8,10,12,14,16,18,20,22,24,25,26,27]]\n",
    "df_filter_sleb = df_sleb[((df_sleb['fold_change'] > 2) & (df_sleb['TPM_ef_mean'] > 5) ) & (~df_sleb['Gene Name'].str.contains('LOC'))]\n",
    "df_filter_sleb.to_csv('sleb_ef5_fold2_noref.csv')\n",
    "\n",
    "gtf_file_path_sleb = 'GCF_003285725.1_SlebRS2_genomic.gff'\n",
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "ref_df_sleb = pd.read_csv(gtf_file_path_sleb, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)\n",
    "ref_gtf_sleb = ref_df_sleb[ref_df_sleb['feature']=='exon']\n",
    "\n",
    "test_sleb = df_sleb.loc[df_filter_sleb.index,:].sort_values(by='Start', ascending=True)\n",
    "added_df_in_sleb = add_overlap_info(test_sleb, ref_gtf_sleb)\n",
    "\n",
    "# filter for df_filter and no overlap regions with existing exons\n",
    "added_df_in_sleb_nooverlap = added_df_in_sleb[added_df_in_sleb['if_overlap'] == 'No'].sort_values(by='fold_change', ascending=False)\n",
    "\n",
    "# .iloc[:,[0,8,10,12,14,16,18,20,22,24,26,27]]\n",
    "added_df_in_sleb_nooverlap.to_csv('sleb_ef5_fold2_noref_nooverlap.csv')\n",
    "\n",
    "transcript_file_path = 'Sleb_m_transcript.fa'\n",
    "df_filter_list = added_df_in_sleb_nooverlap['Gene ID'].to_list()\n",
    "output_file_path = 'sleb_ef5_fold2_noref_nooverlap_seq.fasta'\n",
    "extract_sequences(transcript_file_path, df_filter_list, output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "66b461dd-b79a-483e-8737-a27fc8dd41b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_file_path = 'ef_merged_expression/Sleb_comb.txt'\n",
    "df = pd.read_csv(expression_file_path, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9679cc77-ed35-4f02-b7d2-f41ae9f61d80",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_41839/3275195340.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_gt1['fold_change'] = df_gt1['TPM_ef_mean'] / df_gt1['TPM_ut_mean']\n"
     ]
    }
   ],
   "source": [
    "## Dvir\n",
    "expression_file_path_dvir = 'ef_merged_expression/Dvir_comb.txt'\n",
    "gtf_file_path_dvir = 'ef_merged_gtf/Dvir_merged.gtf'\n",
    "df_dvir = run(expression_file_path_dvir, gtf_file_path_dvir, 2)\n",
    "\n",
    "# filter for foldchange > 2, meanTPM of ef > 5, and don't contain LOC ref gene_id\n",
    "# .iloc[:,[0,4,5,8,10,12,14,16,18,20,22,24,25,26,27]]\n",
    "df_filter_dvir = df_dvir[((df_dvir['fold_change'] > 2) & (df_dvir['TPM_ef_mean'] > 5) ) & (~df_dvir['Gene Name'].str.contains('LOC'))]\n",
    "df_filter_dvir.to_csv('dvir_ef5_fold2_noref.csv')\n",
    "\n",
    "gtf_file_path_dvir = 'GCF_003285735.1_DvirRS2_genomic.gff'\n",
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "ref_df_dvir = pd.read_csv(gtf_file_path_dvir, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)\n",
    "ref_gtf_dvir = ref_df_dvir[ref_df_dvir['feature']=='exon']\n",
    "\n",
    "test_dvir = df_dvir.loc[df_filter_dvir.index,:].sort_values(by='Start', ascending=True)\n",
    "added_df_in_dvir = add_overlap_info(test_dvir, ref_gtf_dvir)\n",
    "\n",
    "# filter for df_filter and no overlap regions with existing exons\n",
    "added_df_in_dvir_nooverlap = added_df_in_dvir[added_df_in_dvir['if_overlap'] == 'No'].sort_values(by='fold_change', ascending=False)\n",
    "\n",
    "# .iloc[:,[0,8,10,12,14,16,18,20,22,24,26,27]]\n",
    "added_df_in_dvir_nooverlap.to_csv('dvir_ef5_fold2_noref_nooverlap.csv')\n",
    "\n",
    "transcript_file_path = 'Dvir_m_transcript.fa'\n",
    "df_filter_list = added_df_in_dvir_nooverlap['Gene ID'].to_list()\n",
    "output_file_path = 'dvir_ef5_fold2_noref_nooverlap_seq.fasta'\n",
    "extract_sequences(transcript_file_path, df_filter_list, output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8bdc6738-736a-448c-a8ed-16a9bcb14602",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_41839/3275195340.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_gt1['fold_change'] = df_gt1['TPM_ef_mean'] / df_gt1['TPM_ut_mean']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(26, 28) (4, 30)\n"
     ]
    }
   ],
   "source": [
    "## Dana\n",
    "expression_file_path_dana = 'ef_merged_expression/Dana_comb.txt'\n",
    "gtf_file_path_dana = 'ef_merged_gtf/Dana_merged.gtf'\n",
    "df_dana = run(expression_file_path_dana, gtf_file_path_dana, 2)\n",
    "\n",
    "# filter for foldchange > 2, meanTPM of ef > 5, and don't contain LOC ref gene_id\n",
    "# .iloc[:,[0,4,5,8,10,12,14,16,18,20,22,24,25,26,27]]\n",
    "df_filter_dana = df_dana[((df_dana['fold_change'] > 2) & (df_dana['TPM_ef_mean'] > 5) ) & (~df_dana['Gene Name'].str.contains('LOC'))]\n",
    "df_filter_dana.to_csv('dana_ef5_fold2_noref.csv')\n",
    "\n",
    "gtf_file_path_dana = 'GCF_017639315.1_ASM1763931v2_genomic.gff'\n",
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "ref_df_dana = pd.read_csv(gtf_file_path_dana, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)\n",
    "ref_gtf_dana = ref_df_dana[ref_df_dana['feature']=='exon']\n",
    "\n",
    "test_dana = df_dana.loc[df_filter_dana.index,:].sort_values(by='Start', ascending=True)\n",
    "added_df_in_dana = add_overlap_info(test_dana, ref_gtf_dana)\n",
    "\n",
    "# filter for df_filter and no overlap regions with existing exons\n",
    "added_df_in_dana_nooverlap = added_df_in_dana[added_df_in_dana['if_overlap'] == 'No'].sort_values(by='fold_change', ascending=False)\n",
    "\n",
    "# .iloc[:,[0,8,10,12,14,16,18,20,22,24,26,27]]\n",
    "added_df_in_dana_nooverlap.to_csv('dana_ef5_fold2_noref_nooverlap.csv')\n",
    "\n",
    "print(df_filter_dana.shape, added_df_in_dana_nooverlap.shape)\n",
    "\n",
    "transcript_file_path = 'Dana_m_transcript.fa'\n",
    "df_filter_list = added_df_in_dana_nooverlap['Gene ID'].to_list()\n",
    "output_file_path = 'dana_ef5_fold2_noref_nooverlap_seq.fasta'\n",
    "extract_sequences(transcript_file_path, df_filter_list, output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "953fbf5d-ea37-4def-9864-e0015eaa2d5b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_41839/3275195340.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_gt1['fold_change'] = df_gt1['TPM_ef_mean'] / df_gt1['TPM_ut_mean']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32, 28) (13, 30)\n"
     ]
    }
   ],
   "source": [
    "## dsim\n",
    "expression_file_path_dsim = 'ef_merged_expression/Dsim_comb.txt'\n",
    "gtf_file_path_dsim = 'ef_merged_gtf/Dsim_merged.gtf'\n",
    "df_dsim = run(expression_file_path_dsim, gtf_file_path_dsim, 2)\n",
    "\n",
    "# filter for foldchange > 2, meanTPM of ef > 5, and don't contain LOC ref gene_id\n",
    "# .iloc[:,[0,4,5,8,10,12,14,16,18,20,22,24,25,26,27]]\n",
    "df_filter_dsim = df_dsim[((df_dsim['fold_change'] > 2) & (df_dsim['TPM_ef_mean'] > 5) ) & (~df_dsim['Gene Name'].str.contains('LOC'))]\n",
    "df_filter_dsim.to_csv('dsim_ef5_fold2_noref.csv')\n",
    "\n",
    "gtf_file_path_dsim = 'GCF_016746395.2_Prin_Dsim_3.1_genomic.gff'\n",
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "ref_df_dsim = pd.read_csv(gtf_file_path_dsim, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)\n",
    "ref_gtf_dsim = ref_df_dsim[ref_df_dsim['feature']=='exon']\n",
    "\n",
    "test_dsim = df_dsim.loc[df_filter_dsim.index,:].sort_values(by='Start', ascending=True)\n",
    "added_df_in_dsim = add_overlap_info(test_dsim, ref_gtf_dsim)\n",
    "\n",
    "# filter for df_filter and no overlap regions with existing exons\n",
    "added_df_in_dsim_nooverlap = added_df_in_dsim[added_df_in_dsim['if_overlap'] == 'No'].sort_values(by='fold_change', ascending=False)\n",
    "\n",
    "# .iloc[:,[0,8,10,12,14,16,18,20,22,24,26,27]]\n",
    "added_df_in_dsim_nooverlap.to_csv('dsim_ef5_fold2_noref_nooverlap.csv')\n",
    "\n",
    "print(df_filter_dsim.shape, added_df_in_dsim_nooverlap.shape)\n",
    "\n",
    "transcript_file_path = 'Dsim_m_transcript.fa'\n",
    "df_filter_list = added_df_in_dsim_nooverlap['Gene ID'].to_list()\n",
    "output_file_path = 'dsim_ef5_fold2_noref_nooverlap_seq.fasta'\n",
    "extract_sequences(transcript_file_path, df_filter_list, output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "98745a07-05b0-4122-a943-80dee032d0b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/jt/wf6lflrs1_94wjjrq_r5mrvm0000gn/T/ipykernel_41839/3275195340.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_gt1['fold_change'] = df_gt1['TPM_ef_mean'] / df_gt1['TPM_ut_mean']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(88, 28) (34, 30)\n"
     ]
    }
   ],
   "source": [
    "## dmel\n",
    "expression_file_path_dmel = 'ef_merged_expression/Dmel_comb.txt'\n",
    "gtf_file_path_dmel = 'ef_merged_gtf/Dmel_merged.gtf'\n",
    "df_dmel = run(expression_file_path_dmel, gtf_file_path_dmel, 2)\n",
    "\n",
    "# filter for foldchange > 2, meanTPM of ef > 5, and don't contain LOC ref gene_id\n",
    "# .iloc[:,[0,4,5,8,10,12,14,16,18,20,22,24,25,26,27]]\n",
    "df_filter_dmel = df_dmel[((df_dmel['fold_change'] > 2) & (df_dmel['TPM_ef_mean'] > 5) ) & (~df_dmel['Gene Name'].str.contains('FBgn'))]\n",
    "df_filter_dmel.to_csv('dmel_ef5_fold2_noref.csv')\n",
    "\n",
    "gtf_file_path_dmel = 'dmel-all-r6.48.gtf'\n",
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "ref_df_dmel = pd.read_csv(gtf_file_path_dmel, sep='\\t', comment='#', header=None, names=column_names, low_memory=False)\n",
    "ref_gtf_dmel = ref_df_dmel[ref_df_dmel['feature']=='exon']\n",
    "\n",
    "\n",
    "test_dmel = df_dmel.loc[df_filter_dmel.index,:].sort_values(by='Start', ascending=True)\n",
    "added_df_in_dmel = add_overlap_info(test_dmel, ref_gtf_dmel)\n",
    "\n",
    "# filter for df_filter and no overlap regions with existing exons\n",
    "added_df_in_dmel_nooverlap = added_df_in_dmel[added_df_in_dmel['if_overlap'] == 'No'].sort_values(by='fold_change', ascending=False)\n",
    "\n",
    "# .iloc[:,[0,8,10,12,14,16,18,20,22,24,26,27]]\n",
    "added_df_in_dmel_nooverlap.to_csv('dmel_ef5_fold2_noref_nooverlap.csv')\n",
    "\n",
    "print(df_filter_dmel.shape, added_df_in_dmel_nooverlap.shape)\n",
    "\n",
    "transcript_file_path = 'Dmel_m_transcript.fa'\n",
    "df_filter_list = added_df_in_dmel_nooverlap['Gene ID'].to_list()\n",
    "output_file_path = 'dmel_ef5_fold2_noref_nooverlap_seq.fasta'\n",
    "extract_sequences(transcript_file_path, df_filter_list, output_file_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
