{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ec51a343-7212-409c-bb84-23274bb00327",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# convert fasta file to dataframe\n",
    "def convert_fasta_to_df(fasta_file):\n",
    "    names = []\n",
    "    descriptions = []\n",
    "    sequences = []\n",
    "    \n",
    "    for record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        names.append(record.id)\n",
    "        descriptions.append(record.description)\n",
    "        sequences.append(str(record.seq))\n",
    "    \n",
    "    dataframe = pd.DataFrame({\n",
    "        'name': names,\n",
    "        'description': descriptions,\n",
    "        'sequence': sequences\n",
    "    })\n",
    "    return dataframe\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# convert a dataframe with columns['name', 'sequence'] to fasta file\n",
    "# specify dataframe and output file path/name as parameters\n",
    "def convert_df_to_fasta(dataframe, output_file):\n",
    "    records = [SeqRecord(Seq(row['sequence']), id=row['name'], description='') for index, row in dataframe.iterrows()]\n",
    "    \n",
    "    with open(output_file, \"w\") as fasta_file:\n",
    "        SeqIO.write(records, fasta_file, \"fasta\")\n",
    "\n",
    "def check_chop_signalp(signalp_gff, df_to_chop, id_column, sequence_column):\n",
    "    # read the signalp output file\n",
    "    columns=['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attribute']\n",
    "    signal = pd.read_csv(signalp_gff, sep='\\t', comment='#', names=columns)\n",
    "    signal = signal.loc[:,['seqid', 'start', 'end', 'score']]\n",
    "    signal['seqid'] = signal['seqid'].str.split().str[0]\n",
    "    # chop the signal peptide off and update the length information\n",
    "    df_chopped = pd.merge(df_to_chop, signal, left_on=id_column, right_on='seqid', how='left')\n",
    "\n",
    "    # fill the default start with 0 for those without signalp\n",
    "    df_chopped['start'].fillna(0, inplace=True)\n",
    "\n",
    "    # chop the signalp off from original sequences\n",
    "    if df_chopped['start'].isin([0,1]).all():\n",
    "        df_chopped['end'] = df_chopped['end'].astype('Int64')\n",
    "        df_chopped['end'].fillna(0, inplace=True)\n",
    "        df_chopped.loc[df_chopped['seqid'].notnull(), sequence_column+'_wo_sp'] = df_chopped.apply(lambda row: row[sequence_column][row['end']:], axis=1)\n",
    "    else:\n",
    "        print('Not all starts at pos 1')\n",
    "        \n",
    "    # update the sequence length to 'length_new'\n",
    "    df_chopped.loc[df_chopped['seqid'].notnull(),'length_wo_sp'] = df_chopped['length'] - df_chopped['end']\n",
    "    df_chopped.drop(columns=['seqid', 'start', 'end', 'score'], inplace=True)\n",
    "    \n",
    "    return df_chopped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b71612fe-564d-44ef-9b37-1956ae34bd56",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# convert orf fasta file from emboss getorf program to dataframe\n",
    "# get orf by $getorf -sequence sleb_pr5_fold2_noref_nooverlap_seq.fasta -find 1 -minsize 15 -outseq test.txt\n",
    "# /rugpfs/fs0/zhao_lab/scratch/cli02/infection_pr/analysis/assembly/merged/orf\n",
    "orf = convert_fasta_to_df('orf/orf_sleb.txt')\n",
    "\n",
    "# add some columns\n",
    "orf['transcript_id'] = orf['name'].str.split('_').str[0]\n",
    "orf['gene_id'] = orf['transcript_id'].apply(lambda x: x.rsplit('.', 1)[0])\n",
    "orf['negative_strand'] = orf['description'].str.contains('REVERSE SENSE')\n",
    "orf['length'] = orf['sequence'].str.len()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "b4a417aa-3a80-43f9-89fb-30408e50c274",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>negative_strand</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.406.1_1</td>\n",
       "      <td>MSTRG.406.1_1 [16 - 42]</td>\n",
       "      <td>MYKHTYIYT</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.406.1_2</td>\n",
       "      <td>MSTRG.406.1_2 [68 - 184]</td>\n",
       "      <td>MCPICFTTKVLDAGEFLRLNCFEQSVLEAKRMQTAETSS</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.406.1_3</td>\n",
       "      <td>MSTRG.406.1_3 [314 - 361]</td>\n",
       "      <td>MPRRREQHTCALPTCQ</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.406.1_4</td>\n",
       "      <td>MSTRG.406.1_4 [355 - 423]</td>\n",
       "      <td>MPVRVILVYYALWSQNWLAGITM</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.406.1_5</td>\n",
       "      <td>MSTRG.406.1_5 [520 - 585]</td>\n",
       "      <td>MLYRRARYKILVDYLLISFIGH</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>MSTRG.13746.1_181</td>\n",
       "      <td>MSTRG.13746.1_181 [352 - 278] (REVERSE SENSE)</td>\n",
       "      <td>MKFFLNNYFKIFFYFLVRHGSLFSR</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>MSTRG.13746.1_182</td>\n",
       "      <td>MSTRG.13746.1_182 [297 - 241] (REVERSE SENSE)</td>\n",
       "      <td>MVLSFQDDSRVTEVVTLLM</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>MSTRG.13746.1_183</td>\n",
       "      <td>MSTRG.13746.1_183 [213 - 184] (REVERSE SENSE)</td>\n",
       "      <td>MYLYILFGRI</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>MSTRG.13746.1_184</td>\n",
       "      <td>MSTRG.13746.1_184 [238 - 140] (REVERSE SENSE)</td>\n",
       "      <td>MISEGCNMYVFIYSFWSDLIAKDTKLMGKSFKS</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>MSTRG.13746.1_185</td>\n",
       "      <td>MSTRG.13746.1_185 [78 - 1] (REVERSE SENSE)</td>\n",
       "      <td>MFPRIKYQVSCFHNFKFNKLFRTSRQ</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>523 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  name                                    description  \\\n",
       "0        MSTRG.406.1_1                        MSTRG.406.1_1 [16 - 42]   \n",
       "1        MSTRG.406.1_2                       MSTRG.406.1_2 [68 - 184]   \n",
       "2        MSTRG.406.1_3                      MSTRG.406.1_3 [314 - 361]   \n",
       "3        MSTRG.406.1_4                      MSTRG.406.1_4 [355 - 423]   \n",
       "4        MSTRG.406.1_5                      MSTRG.406.1_5 [520 - 585]   \n",
       "..                 ...                                            ...   \n",
       "518  MSTRG.13746.1_181  MSTRG.13746.1_181 [352 - 278] (REVERSE SENSE)   \n",
       "519  MSTRG.13746.1_182  MSTRG.13746.1_182 [297 - 241] (REVERSE SENSE)   \n",
       "520  MSTRG.13746.1_183  MSTRG.13746.1_183 [213 - 184] (REVERSE SENSE)   \n",
       "521  MSTRG.13746.1_184  MSTRG.13746.1_184 [238 - 140] (REVERSE SENSE)   \n",
       "522  MSTRG.13746.1_185     MSTRG.13746.1_185 [78 - 1] (REVERSE SENSE)   \n",
       "\n",
       "                                    sequence  transcript_id      gene_id  \\\n",
       "0                                  MYKHTYIYT    MSTRG.406.1    MSTRG.406   \n",
       "1    MCPICFTTKVLDAGEFLRLNCFEQSVLEAKRMQTAETSS    MSTRG.406.1    MSTRG.406   \n",
       "2                           MPRRREQHTCALPTCQ    MSTRG.406.1    MSTRG.406   \n",
       "3                    MPVRVILVYYALWSQNWLAGITM    MSTRG.406.1    MSTRG.406   \n",
       "4                     MLYRRARYKILVDYLLISFIGH    MSTRG.406.1    MSTRG.406   \n",
       "..                                       ...            ...          ...   \n",
       "518                MKFFLNNYFKIFFYFLVRHGSLFSR  MSTRG.13746.1  MSTRG.13746   \n",
       "519                      MVLSFQDDSRVTEVVTLLM  MSTRG.13746.1  MSTRG.13746   \n",
       "520                               MYLYILFGRI  MSTRG.13746.1  MSTRG.13746   \n",
       "521        MISEGCNMYVFIYSFWSDLIAKDTKLMGKSFKS  MSTRG.13746.1  MSTRG.13746   \n",
       "522               MFPRIKYQVSCFHNFKFNKLFRTSRQ  MSTRG.13746.1  MSTRG.13746   \n",
       "\n",
       "     negative_strand  length  \n",
       "0              False       9  \n",
       "1              False      39  \n",
       "2              False      16  \n",
       "3              False      23  \n",
       "4              False      22  \n",
       "..               ...     ...  \n",
       "518             True      25  \n",
       "519             True      19  \n",
       "520             True      10  \n",
       "521             True      33  \n",
       "522             True      26  \n",
       "\n",
       "[523 rows x 7 columns]"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f00ecd37-c465-444d-b5a0-25b78e93d474",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check and chop signal peptides from signalp6 results\n",
    "# $ signalp6 --fastafile orf_sleb.txt --output_dir signalp_sleb --organism eukarya --mode fast \n",
    "# rugpfs/fs0/zhao_lab/scratch/cli02/infection_pr/analysis/assembly/merged/orf/signalp_sleb\n",
    "orf = check_chop_signalp('orf/output.gff3', orf, 'name', 'sequence')\n",
    "\n",
    "orf['signal_peptide'] = ~orf['length_new'].isnull()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "75985100-fde9-4b47-b408-5d84810b9f12",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>negative_strand</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence_new</th>\n",
       "      <th>length_new</th>\n",
       "      <th>signal_peptide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.406.1_1</td>\n",
       "      <td>MSTRG.406.1_1 [16 - 42]</td>\n",
       "      <td>MYKHTYIYT</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.406.1_2</td>\n",
       "      <td>MSTRG.406.1_2 [68 - 184]</td>\n",
       "      <td>MCPICFTTKVLDAGEFLRLNCFEQSVLEAKRMQTAETSS</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.406.1_3</td>\n",
       "      <td>MSTRG.406.1_3 [314 - 361]</td>\n",
       "      <td>MPRRREQHTCALPTCQ</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.406.1_4</td>\n",
       "      <td>MSTRG.406.1_4 [355 - 423]</td>\n",
       "      <td>MPVRVILVYYALWSQNWLAGITM</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.406.1_5</td>\n",
       "      <td>MSTRG.406.1_5 [520 - 585]</td>\n",
       "      <td>MLYRRARYKILVDYLLISFIGH</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MSTRG.406.1_6</td>\n",
       "      <td>MSTRG.406.1_6 [661 - 717]</td>\n",
       "      <td>MQILTSMPLFWTFSTARLC</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MSTRG.406.1_7</td>\n",
       "      <td>MSTRG.406.1_7 [730 - 756]</td>\n",
       "      <td>MLEIDPQLG</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MSTRG.406.1_8</td>\n",
       "      <td>MSTRG.406.1_8 [928 - 1122]</td>\n",
       "      <td>MEPNSNLSVNILNSPRPHLYRNCRLIWPYCCLTTMCRFLPPGHYFN...</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MSTRG.406.1_9</td>\n",
       "      <td>MSTRG.406.1_9 [1115 - 1207]</td>\n",
       "      <td>MEHKTVHGGISDRRHLHAARRSAALWSGTKG</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MSTRG.406.1_10</td>\n",
       "      <td>MSTRG.406.1_10 [1318 - 1371]</td>\n",
       "      <td>MPCLRLIYWRIRTYVHLS</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MSTRG.406.1_11</td>\n",
       "      <td>MSTRG.406.1_11 [1414 - 1467]</td>\n",
       "      <td>MCQSWVCQFLETKEGTCK</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MSTRG.406.1_12</td>\n",
       "      <td>MSTRG.406.1_12 [1828 - 1848]</td>\n",
       "      <td>MPKKDPN</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MSTRG.406.1_13</td>\n",
       "      <td>MSTRG.406.1_13 [240 - 1853]</td>\n",
       "      <td>MHIEMGFGTMPGRIVLFCLLLISNICRGEGSNILVLFPHASESHFG...</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>538</td>\n",
       "      <td>EGSNILVLFPHASESHFGVLRTLVTELASRDHNVTVYSTHGLGETL...</td>\n",
       "      <td>510</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MSTRG.406.1_14</td>\n",
       "      <td>MSTRG.406.1_14 [1903 - 1920]</td>\n",
       "      <td>MLRNLK</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MSTRG.406.1_15</td>\n",
       "      <td>MSTRG.406.1_15 [1738 - 1664] (REVERSE SENSE)</td>\n",
       "      <td>MKRSKRCSWQEEIPTPRVLIALPPL</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>True</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              name                                   description  \\\n",
       "0    MSTRG.406.1_1                       MSTRG.406.1_1 [16 - 42]   \n",
       "1    MSTRG.406.1_2                      MSTRG.406.1_2 [68 - 184]   \n",
       "2    MSTRG.406.1_3                     MSTRG.406.1_3 [314 - 361]   \n",
       "3    MSTRG.406.1_4                     MSTRG.406.1_4 [355 - 423]   \n",
       "4    MSTRG.406.1_5                     MSTRG.406.1_5 [520 - 585]   \n",
       "5    MSTRG.406.1_6                     MSTRG.406.1_6 [661 - 717]   \n",
       "6    MSTRG.406.1_7                     MSTRG.406.1_7 [730 - 756]   \n",
       "7    MSTRG.406.1_8                    MSTRG.406.1_8 [928 - 1122]   \n",
       "8    MSTRG.406.1_9                   MSTRG.406.1_9 [1115 - 1207]   \n",
       "9   MSTRG.406.1_10                  MSTRG.406.1_10 [1318 - 1371]   \n",
       "10  MSTRG.406.1_11                  MSTRG.406.1_11 [1414 - 1467]   \n",
       "11  MSTRG.406.1_12                  MSTRG.406.1_12 [1828 - 1848]   \n",
       "12  MSTRG.406.1_13                   MSTRG.406.1_13 [240 - 1853]   \n",
       "13  MSTRG.406.1_14                  MSTRG.406.1_14 [1903 - 1920]   \n",
       "14  MSTRG.406.1_15  MSTRG.406.1_15 [1738 - 1664] (REVERSE SENSE)   \n",
       "\n",
       "                                             sequence transcript_id  \\\n",
       "0                                           MYKHTYIYT   MSTRG.406.1   \n",
       "1             MCPICFTTKVLDAGEFLRLNCFEQSVLEAKRMQTAETSS   MSTRG.406.1   \n",
       "2                                    MPRRREQHTCALPTCQ   MSTRG.406.1   \n",
       "3                             MPVRVILVYYALWSQNWLAGITM   MSTRG.406.1   \n",
       "4                              MLYRRARYKILVDYLLISFIGH   MSTRG.406.1   \n",
       "5                                 MQILTSMPLFWTFSTARLC   MSTRG.406.1   \n",
       "6                                           MLEIDPQLG   MSTRG.406.1   \n",
       "7   MEPNSNLSVNILNSPRPHLYRNCRLIWPYCCLTTMCRFLPPGHYFN...   MSTRG.406.1   \n",
       "8                     MEHKTVHGGISDRRHLHAARRSAALWSGTKG   MSTRG.406.1   \n",
       "9                                  MPCLRLIYWRIRTYVHLS   MSTRG.406.1   \n",
       "10                                 MCQSWVCQFLETKEGTCK   MSTRG.406.1   \n",
       "11                                            MPKKDPN   MSTRG.406.1   \n",
       "12  MHIEMGFGTMPGRIVLFCLLLISNICRGEGSNILVLFPHASESHFG...   MSTRG.406.1   \n",
       "13                                             MLRNLK   MSTRG.406.1   \n",
       "14                          MKRSKRCSWQEEIPTPRVLIALPPL   MSTRG.406.1   \n",
       "\n",
       "      gene_id  negative_strand  length  \\\n",
       "0   MSTRG.406            False       9   \n",
       "1   MSTRG.406            False      39   \n",
       "2   MSTRG.406            False      16   \n",
       "3   MSTRG.406            False      23   \n",
       "4   MSTRG.406            False      22   \n",
       "5   MSTRG.406            False      19   \n",
       "6   MSTRG.406            False       9   \n",
       "7   MSTRG.406            False      65   \n",
       "8   MSTRG.406            False      31   \n",
       "9   MSTRG.406            False      18   \n",
       "10  MSTRG.406            False      18   \n",
       "11  MSTRG.406            False       7   \n",
       "12  MSTRG.406            False     538   \n",
       "13  MSTRG.406            False       6   \n",
       "14  MSTRG.406             True      25   \n",
       "\n",
       "                                         sequence_new  length_new  \\\n",
       "0                                                 NaN        <NA>   \n",
       "1                                                 NaN        <NA>   \n",
       "2                                                 NaN        <NA>   \n",
       "3                                                 NaN        <NA>   \n",
       "4                                                 NaN        <NA>   \n",
       "5                                                 NaN        <NA>   \n",
       "6                                                 NaN        <NA>   \n",
       "7                                                 NaN        <NA>   \n",
       "8                                                 NaN        <NA>   \n",
       "9                                                 NaN        <NA>   \n",
       "10                                                NaN        <NA>   \n",
       "11                                                NaN        <NA>   \n",
       "12  EGSNILVLFPHASESHFGVLRTLVTELASRDHNVTVYSTHGLGETL...         510   \n",
       "13                                                NaN        <NA>   \n",
       "14                                                NaN        <NA>   \n",
       "\n",
       "    signal_peptide  \n",
       "0            False  \n",
       "1            False  \n",
       "2            False  \n",
       "3            False  \n",
       "4            False  \n",
       "5            False  \n",
       "6            False  \n",
       "7            False  \n",
       "8            False  \n",
       "9            False  \n",
       "10           False  \n",
       "11           False  \n",
       "12            True  \n",
       "13           False  \n",
       "14           False  "
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "7d59c760-c5c0-4d0a-8e04-456d6483bebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the information dataframe containing the reference strand info in gtf\n",
    "info = pd.read_csv('sleb_pr5_fold2_noref_nooverlap.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "09c93bd2-15c9-49fb-b369-6c2baf3df73d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the info the orf dataframe\n",
    "orf = orf.merge(info[['Gene ID', 'Strand']], how='left', left_on='gene_id', right_on='Gene ID')\n",
    "orf.drop(columns=['Gene ID'], inplace=True)\n",
    "\n",
    "# remove the orfs that are on the wrong strand\n",
    "orf = orf[~((orf['negative_strand']==False) & (orf['Strand']=='-'))]\n",
    "orf = orf[~((orf['negative_strand']==True) & (orf['Strand']=='+'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "96998095-931a-4266-933f-d0c5b9a0e43a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>negative_strand</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence_new</th>\n",
       "      <th>length_new</th>\n",
       "      <th>signal_peptide</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.406.1_1</td>\n",
       "      <td>MSTRG.406.1_1 [16 - 42]</td>\n",
       "      <td>MYKHTYIYT</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.406.1_2</td>\n",
       "      <td>MSTRG.406.1_2 [68 - 184]</td>\n",
       "      <td>MCPICFTTKVLDAGEFLRLNCFEQSVLEAKRMQTAETSS</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSTRG.406.1_3</td>\n",
       "      <td>MSTRG.406.1_3 [314 - 361]</td>\n",
       "      <td>MPRRREQHTCALPTCQ</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSTRG.406.1_4</td>\n",
       "      <td>MSTRG.406.1_4 [355 - 423]</td>\n",
       "      <td>MPVRVILVYYALWSQNWLAGITM</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.406.1_5</td>\n",
       "      <td>MSTRG.406.1_5 [520 - 585]</td>\n",
       "      <td>MLYRRARYKILVDYLLISFIGH</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>MSTRG.13746.1_181</td>\n",
       "      <td>MSTRG.13746.1_181 [352 - 278] (REVERSE SENSE)</td>\n",
       "      <td>MKFFLNNYFKIFFYFLVRHGSLFSR</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>MSTRG.13746.1_182</td>\n",
       "      <td>MSTRG.13746.1_182 [297 - 241] (REVERSE SENSE)</td>\n",
       "      <td>MVLSFQDDSRVTEVVTLLM</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>MSTRG.13746.1_183</td>\n",
       "      <td>MSTRG.13746.1_183 [213 - 184] (REVERSE SENSE)</td>\n",
       "      <td>MYLYILFGRI</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>MSTRG.13746.1_184</td>\n",
       "      <td>MSTRG.13746.1_184 [238 - 140] (REVERSE SENSE)</td>\n",
       "      <td>MISEGCNMYVFIYSFWSDLIAKDTKLMGKSFKS</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>MSTRG.13746.1_185</td>\n",
       "      <td>MSTRG.13746.1_185 [78 - 1] (REVERSE SENSE)</td>\n",
       "      <td>MFPRIKYQVSCFHNFKFNKLFRTSRQ</td>\n",
       "      <td>MSTRG.13746.1</td>\n",
       "      <td>MSTRG.13746</td>\n",
       "      <td>True</td>\n",
       "      <td>26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>336 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  name                                    description  \\\n",
       "0        MSTRG.406.1_1                        MSTRG.406.1_1 [16 - 42]   \n",
       "1        MSTRG.406.1_2                       MSTRG.406.1_2 [68 - 184]   \n",
       "2        MSTRG.406.1_3                      MSTRG.406.1_3 [314 - 361]   \n",
       "3        MSTRG.406.1_4                      MSTRG.406.1_4 [355 - 423]   \n",
       "4        MSTRG.406.1_5                      MSTRG.406.1_5 [520 - 585]   \n",
       "..                 ...                                            ...   \n",
       "518  MSTRG.13746.1_181  MSTRG.13746.1_181 [352 - 278] (REVERSE SENSE)   \n",
       "519  MSTRG.13746.1_182  MSTRG.13746.1_182 [297 - 241] (REVERSE SENSE)   \n",
       "520  MSTRG.13746.1_183  MSTRG.13746.1_183 [213 - 184] (REVERSE SENSE)   \n",
       "521  MSTRG.13746.1_184  MSTRG.13746.1_184 [238 - 140] (REVERSE SENSE)   \n",
       "522  MSTRG.13746.1_185     MSTRG.13746.1_185 [78 - 1] (REVERSE SENSE)   \n",
       "\n",
       "                                    sequence  transcript_id      gene_id  \\\n",
       "0                                  MYKHTYIYT    MSTRG.406.1    MSTRG.406   \n",
       "1    MCPICFTTKVLDAGEFLRLNCFEQSVLEAKRMQTAETSS    MSTRG.406.1    MSTRG.406   \n",
       "2                           MPRRREQHTCALPTCQ    MSTRG.406.1    MSTRG.406   \n",
       "3                    MPVRVILVYYALWSQNWLAGITM    MSTRG.406.1    MSTRG.406   \n",
       "4                     MLYRRARYKILVDYLLISFIGH    MSTRG.406.1    MSTRG.406   \n",
       "..                                       ...            ...          ...   \n",
       "518                MKFFLNNYFKIFFYFLVRHGSLFSR  MSTRG.13746.1  MSTRG.13746   \n",
       "519                      MVLSFQDDSRVTEVVTLLM  MSTRG.13746.1  MSTRG.13746   \n",
       "520                               MYLYILFGRI  MSTRG.13746.1  MSTRG.13746   \n",
       "521        MISEGCNMYVFIYSFWSDLIAKDTKLMGKSFKS  MSTRG.13746.1  MSTRG.13746   \n",
       "522               MFPRIKYQVSCFHNFKFNKLFRTSRQ  MSTRG.13746.1  MSTRG.13746   \n",
       "\n",
       "     negative_strand  length sequence_new  length_new  signal_peptide Strand  \n",
       "0              False       9          NaN        <NA>           False      +  \n",
       "1              False      39          NaN        <NA>           False      +  \n",
       "2              False      16          NaN        <NA>           False      +  \n",
       "3              False      23          NaN        <NA>           False      +  \n",
       "4              False      22          NaN        <NA>           False      +  \n",
       "..               ...     ...          ...         ...             ...    ...  \n",
       "518             True      25          NaN        <NA>           False      -  \n",
       "519             True      19          NaN        <NA>           False      -  \n",
       "520             True      10          NaN        <NA>           False      -  \n",
       "521             True      33          NaN        <NA>           False      -  \n",
       "522             True      26          NaN        <NA>           False      -  \n",
       "\n",
       "[336 rows x 11 columns]"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "f73fc989-06d9-4225-91d1-fead434e3850",
   "metadata": {},
   "outputs": [],
   "source": [
    "# group dataframe for each gene id and each strand\n",
    "orf_groups = orf.groupby(['gene_id', 'negative_strand'])\n",
    "\n",
    "# get the longest orf in each group\n",
    "longest_orf = orf_groups['length'].apply(lambda x: max(x))\n",
    "\n",
    "# Reset the index of longest_sequence to match it with the original DataFrame\n",
    "longest_orf = longest_orf.reset_index()\n",
    "\n",
    "# Merge the longest sequences with the original DataFrame\n",
    "orf_longest_each_strand = pd.merge(orf, longest_orf, on=['gene_id', 'negative_strand', 'length'], how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "cbbd8620-901e-49a9-81c8-b62efd307a07",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# extract candidates with signal peptide\n",
    "candidate = orf_longest_each_strand[orf_longest_each_strand['signal_peptide']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "051295ad-453d-439b-9f4b-14bd9ef0b0a8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>negative_strand</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence_new</th>\n",
       "      <th>length_new</th>\n",
       "      <th>signal_peptide</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSTRG.406.1_13</td>\n",
       "      <td>MSTRG.406.1_13 [240 - 1853]</td>\n",
       "      <td>MHIEMGFGTMPGRIVLFCLLLISNICRGEGSNILVLFPHASESHFG...</td>\n",
       "      <td>MSTRG.406.1</td>\n",
       "      <td>MSTRG.406</td>\n",
       "      <td>False</td>\n",
       "      <td>538</td>\n",
       "      <td>EGSNILVLFPHASESHFGVLRTLVTELASRDHNVTVYSTHGLGETL...</td>\n",
       "      <td>510</td>\n",
       "      <td>True</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSTRG.1588.1_1</td>\n",
       "      <td>MSTRG.1588.1_1 [58 - 210]</td>\n",
       "      <td>MFFNIRTLLLPLLLLGLALTGTVTEASLQRGRVFDTRPSPYNPNPP...</td>\n",
       "      <td>MSTRG.1588.1</td>\n",
       "      <td>MSTRG.1588</td>\n",
       "      <td>False</td>\n",
       "      <td>51</td>\n",
       "      <td>SLQRGRVFDTRPSPYNPNPPRPGPF</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTRG.2269.1_4</td>\n",
       "      <td>MSTRG.2269.1_4 [249 - 76] (REVERSE SENSE)</td>\n",
       "      <td>MRTLSLVYALLMLCFLGFPNLGRAIPFVDRRGGSWGGGQTMRPIPT...</td>\n",
       "      <td>MSTRG.2269.1</td>\n",
       "      <td>MSTRG.2269</td>\n",
       "      <td>True</td>\n",
       "      <td>58</td>\n",
       "      <td>IPFVDRRGGSWGGGQTMRPIPTTPPFNPHASRRF</td>\n",
       "      <td>34</td>\n",
       "      <td>True</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MSTRG.4616.1_6</td>\n",
       "      <td>MSTRG.4616.1_6 [421 - 107] (REVERSE SENSE)</td>\n",
       "      <td>MALKYTLFFLACMLAYVAAEQYEFDGNKVRPLGGNGLPEGGRKVEL...</td>\n",
       "      <td>MSTRG.4616.1</td>\n",
       "      <td>MSTRG.4616</td>\n",
       "      <td>True</td>\n",
       "      <td>105</td>\n",
       "      <td>EQYEFDGNKVRPLGGNGLPEGGRKVELKYTDDPAQGGRQSSVRFDQ...</td>\n",
       "      <td>86</td>\n",
       "      <td>True</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MSTRG.6046.1_4</td>\n",
       "      <td>MSTRG.6046.1_4 [195 - 329]</td>\n",
       "      <td>MMGQVFMHFALLCCCAATAAQDQICAPIRLASGHGNGQRQRPRLV</td>\n",
       "      <td>MSTRG.6046.1</td>\n",
       "      <td>MSTRG.6046</td>\n",
       "      <td>False</td>\n",
       "      <td>45</td>\n",
       "      <td>QDQICAPIRLASGHGNGQRQRPRLV</td>\n",
       "      <td>25</td>\n",
       "      <td>True</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MSTRG.7930.1_10</td>\n",
       "      <td>MSTRG.7930.1_10 [580 - 398] (REVERSE SENSE)</td>\n",
       "      <td>MQFNALCFLLLSLCIVCALNSAVAAVCSGHGFDSRNGNLNLDINNL...</td>\n",
       "      <td>MSTRG.7930.1</td>\n",
       "      <td>MSTRG.7930</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>AVCSGHGFDSRNGNLNLDINNLDSLFRCVEQQHRNRG</td>\n",
       "      <td>37</td>\n",
       "      <td>True</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               name                                  description  \\\n",
       "0    MSTRG.406.1_13                  MSTRG.406.1_13 [240 - 1853]   \n",
       "1    MSTRG.1588.1_1                    MSTRG.1588.1_1 [58 - 210]   \n",
       "4    MSTRG.2269.1_4    MSTRG.2269.1_4 [249 - 76] (REVERSE SENSE)   \n",
       "13   MSTRG.4616.1_6   MSTRG.4616.1_6 [421 - 107] (REVERSE SENSE)   \n",
       "14   MSTRG.6046.1_4                   MSTRG.6046.1_4 [195 - 329]   \n",
       "19  MSTRG.7930.1_10  MSTRG.7930.1_10 [580 - 398] (REVERSE SENSE)   \n",
       "\n",
       "                                             sequence transcript_id  \\\n",
       "0   MHIEMGFGTMPGRIVLFCLLLISNICRGEGSNILVLFPHASESHFG...   MSTRG.406.1   \n",
       "1   MFFNIRTLLLPLLLLGLALTGTVTEASLQRGRVFDTRPSPYNPNPP...  MSTRG.1588.1   \n",
       "4   MRTLSLVYALLMLCFLGFPNLGRAIPFVDRRGGSWGGGQTMRPIPT...  MSTRG.2269.1   \n",
       "13  MALKYTLFFLACMLAYVAAEQYEFDGNKVRPLGGNGLPEGGRKVEL...  MSTRG.4616.1   \n",
       "14      MMGQVFMHFALLCCCAATAAQDQICAPIRLASGHGNGQRQRPRLV  MSTRG.6046.1   \n",
       "19  MQFNALCFLLLSLCIVCALNSAVAAVCSGHGFDSRNGNLNLDINNL...  MSTRG.7930.1   \n",
       "\n",
       "       gene_id  negative_strand  length  \\\n",
       "0    MSTRG.406            False     538   \n",
       "1   MSTRG.1588            False      51   \n",
       "4   MSTRG.2269             True      58   \n",
       "13  MSTRG.4616             True     105   \n",
       "14  MSTRG.6046            False      45   \n",
       "19  MSTRG.7930             True      61   \n",
       "\n",
       "                                         sequence_new  length_new  \\\n",
       "0   EGSNILVLFPHASESHFGVLRTLVTELASRDHNVTVYSTHGLGETL...         510   \n",
       "1                           SLQRGRVFDTRPSPYNPNPPRPGPF          25   \n",
       "4                  IPFVDRRGGSWGGGQTMRPIPTTPPFNPHASRRF          34   \n",
       "13  EQYEFDGNKVRPLGGNGLPEGGRKVELKYTDDPAQGGRQSSVRFDQ...          86   \n",
       "14                          QDQICAPIRLASGHGNGQRQRPRLV          25   \n",
       "19              AVCSGHGFDSRNGNLNLDINNLDSLFRCVEQQHRNRG          37   \n",
       "\n",
       "    signal_peptide Strand  \n",
       "0             True      +  \n",
       "1             True      .  \n",
       "4             True      .  \n",
       "13            True      .  \n",
       "14            True      .  \n",
       "19            True      .  "
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0bd7e01-6144-4e6b-a096-88737b151f96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to csv and fasta files\n",
    "candidate.to_csv('orf/candidate_orf_sleb.csv')\n",
    "convert_df_to_fasta(candidate, 'orf/candidate_orf_sleb.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "553fe327-4202-491e-8a80-de84dd607304",
   "metadata": {},
   "outputs": [],
   "source": [
    "# standardized way to get orf, add signalp info, add strand info, and get candidates\n",
    "# prepare orf fasta file as 'orf/orf_'+species+'.txt', info file species+'_pr5_fold2_noref_nooverlap.csv',/\n",
    "# signalp result gff file 'orf/output.gff3'\n",
    "# function would output candidate info to file 'orf/candidate_orf_'+species+'.csv' and sequence to 'orf/candidate_orf_'+species+'.fasta'\n",
    "# also a transcript with signalp only version would be output as well\n",
    "\n",
    "def process_orf(species): # species name should be hard-encoded species name as 'sleb' as the parameter\n",
    "    # convert orf fasta file from emboss getorf program to dataframe\n",
    "    orf = convert_fasta_to_df('orf/orf_'+species+'.txt')\n",
    "    \n",
    "    # add some columns\n",
    "    orf['transcript_id'] = orf['name'].str.split('_').str[0]\n",
    "    orf['gene_id'] = orf['transcript_id'].apply(lambda x: x.rsplit('.', 1)[0])\n",
    "    orf['negative_strand'] = orf['description'].str.contains('REVERSE SENSE')\n",
    "    orf['length'] = orf['sequence'].str.len()\n",
    "    \n",
    "    # check and chop signal peptides from signalp6 results\n",
    "    orf = check_chop_signalp('orf/'+species+'_output.gff3', orf, 'name', 'sequence')\n",
    "    orf['signal_peptide'] = ~orf['length_wo_sp'].isnull()\n",
    "    \n",
    "    # load the information dataframe containing the reference strand info in gtf\n",
    "    info = pd.read_csv(species+'_ef5_fold2_noref_nooverlap.csv')\n",
    "    info.drop(columns=['Unnamed: 0', 'Gene Name', '%overlap'], inplace=True)\n",
    "    \n",
    "    # add the info the orf dataframe\n",
    "    orf = orf.merge(info[['Gene ID', 'Strand']], how='left', left_on='gene_id', right_on='Gene ID')\n",
    "    orf.drop(columns=['Gene ID'], inplace=True)\n",
    "    \n",
    "    # remove the orfs that are on the wrong strand\n",
    "    orf = orf[~((orf['negative_strand']==False) & (orf['Strand']=='-'))]\n",
    "    orf = orf[~((orf['negative_strand']==True) & (orf['Strand']=='+'))]\n",
    "    \n",
    "    # group dataframe for each gene id and each strand\n",
    "    orf_groups = orf.groupby(['gene_id', 'negative_strand'])\n",
    "    \n",
    "    # get the longest orf in each group\n",
    "    longest_orf = orf_groups['length'].apply(lambda x: max(x))\n",
    "    \n",
    "    # Reset the index of longest_sequence to match it with the original DataFrame\n",
    "    longest_orf = longest_orf.reset_index()\n",
    "    \n",
    "    # Merge the longest sequences with the original DataFrame\n",
    "    orf_longest_each_strand = pd.merge(orf, longest_orf, on=['gene_id', 'negative_strand', 'length'], how='inner')\n",
    "\n",
    "    candidate = orf_longest_each_strand\n",
    "    # extract candidates with signal peptide\n",
    "    candidate_sigp = orf_longest_each_strand[orf_longest_each_strand['signal_peptide']]\n",
    "\n",
    "    info.drop(columns=['Strand'], inplace=True)\n",
    "    combined = candidate.merge(info, how='left', left_on='gene_id', right_on='Gene ID')\n",
    "    combined_sigp = candidate_sigp.merge(info, how='left', left_on='gene_id', right_on='Gene ID')\n",
    "    \n",
    "    # save to csv and fasta files\n",
    "    candidate.to_csv('orf/candidate_orf_'+species+'.csv')\n",
    "    candidate_sigp.to_csv('orf/candidate_orf_sigp_'+species+'.csv')\n",
    "    convert_df_to_fasta(candidate, 'orf/candidate_orf_'+species+'.fasta')\n",
    "    convert_df_to_fasta(candidate_sigp, 'orf/candidate_orf_sigp_'+species+'.fasta')\n",
    "    combined.to_csv('orf/candidate_orf_combined_'+species+'.csv')\n",
    "    combined_sigp.to_csv('orf/candidate_orf_combined_sigp_'+species+'.csv')\n",
    "\n",
    "    print(f'Finished processing orf for {species}.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "568b432b-b659-432c-9f12-21db5e97b5ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished processing orf for dsim.\n",
      "Finished processing orf for dana.\n",
      "Finished processing orf for dvir.\n",
      "Finished processing orf for sleb.\n"
     ]
    }
   ],
   "source": [
    "species_list = ['dsim', 'dana', 'dvir', 'sleb']\n",
    "for species in species_list:\n",
    "    process_orf(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f05d49f6-bae7-4b4d-a5bf-b2afc2b7051c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
